### PR TITLE
feat: sso integration

### LINKED ISSUE #3166: SSO integration
## Description
SSO (Single Sign-On) integration: Many institutions that use CircuitVerse use a local copy of CircuitVerse with a custom SSO authentication. We intend to make this task easier for the institutions by natively supporting SSO with CircuitVerse

### PR DESCRIPTION
**Pull request by vedant-jain03**
Fixes #3166 

#### Describe the changes you have made in this PR -
- [x] added a gem `devise_saml_authenticatable`
- [x] added `attribute-map.yml` (This tells Devise SAML what attributes returned from the Id Provider map to which attribute of your User model.)
- [x] added `devise` configuration
- [x] added route for controller under `user` scope
- [x] added UI

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

File: app/controllers/users/saml_sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  after_action :store_winning_strategy, only: :create
+
+  def create
+    super
+  end
+
+  private
+
+    def store_winning_strategy
+      warden.session(:user)[:strategy] = warden.winning_strategies[:user].class.name.demodulize.underscore.to_sym
```
[the remark this answers is not in the document]
Comment (vedant-jain03) [reply app/controllers/users/saml_sessions_controller.rb:13]: It is needed

File: app/models/user.rb [hunk 2]
```diff
@@ -14,7 +14,7 @@ class User < ApplicationRecord
   has_many :rated_projects, through: :stars, dependent: :destroy, source: "project"
   has_many :groups_mentored, class_name: "Group",  foreign_key: "mentor_id", dependent: :destroy
   devise :database_authenticatable, :registerable, :recoverable, :rememberable, :trackable,
-         :validatable, :omniauthable, omniauth_providers: %i[google_oauth2 facebook github gitlab]
+         :validatable, :omniauthable, :saml_authenticatable, omniauth_providers: %i[google_oauth2 facebook github gitlab]
```
[the remark this answers is not in the document]
Comment (vedant-jain03) [reply app/models/user.rb:17]: Expected

File: app/views/users/shared/_links.html.erb [hunk 3]
```diff
@@ -11,5 +11,11 @@
         <%= link_to image_tag("SVGs/gitlab.svg", :alt=>"Gitlab Icon", :class=>"users-social-links"), user_gitlab_omniauth_authorize_path, method: :post %>
       <% end %>
     </div>
+    <div class="row center-row users-shared-row">
```
Comment (tachyons) [inline app/views/users/shared/_links.html.erb:14]: Keep this under feature flag

File: config/attribute-map.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+    "urn:mace:dir:attribute-def:uid": "uid"
+    "urn:mace:dir:attribute-def:email": "email"
+    "urn:mace:dir:attribute-def:name": "name"
+    "urn:mace:dir:attribute-def:password": "encrypted_password"
```
Comment (tachyons) [inline config/attribute-map.yml:4]: Are you sure that password attribute is needed here?

File: config/initializers/devise.rb [hunk 5]
```diff
@@ -38,6 +38,27 @@
   # :mongoid (bson_ext recommended) by default. Other ORMs may be
   # available as additional gems.
   require 'devise/orm/active_record'
+  require 'securerandom'
+  # saml configuration for devise
+  $callback = Rails.env.development? ? 'http://localhost:3000' : ENV['CALLBACK_ADDRESS']
```
Comment (tachyons) [inline config/initializers/devise.rb:43]: Does this need to be a global variable ?

File: config/attribute-map.yml (hunk 4 above)
Comment (vedant-jain03) [reply config/attribute-map.yml]: using only `uid` is working fine!
Removed other!

**Comment by tachyons** [comment]
Great 👍

**Pull request merged**